pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,430 @@
+package nodelink
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	machineAnnotationKey = "machine.openshift.io/machine"
+)
+
+// blank assignment to verify that ReconcileNodeLink implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNodeLink{}
+
+// ReconcileNodeLink reconciles a Node object
+type ReconcileNodeLink struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	// This is useful for unit testing so we can mock cache IndexField
+	// and emulate Client.List.MatchingField behaviour
+	listNodesByFieldFunc    func(key, value string) ([]corev1.Node, error)
+	listMachinesByFieldFunc func(key, value string) ([]mapiv1beta1.Machine, error)
+}
+
+// Add creates a new Nodelink Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	reconciler, err := newReconciler(mgr)
+	if err != nil {
+		return fmt.Errorf("error building reconciler: %v", err)
+	}
+	return add(mgr, reconciler, reconciler.nodeRequestFromMachine)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) (*ReconcileNodeLink, error) {
+	// set convenient indexers
+	if err := mgr.GetCache().IndexField(&corev1.Node{},
+		"ProviderID",
+		func(object runtime.Object) []string {
```

Comment:
Please name the anonymous functions:
- so they can be unit tested
- to get:

```go
const (
	machineInternalIPIndex = "machineInternalIPIndex"
	machineProviderIDIndex = "machineProviderIDIndex"
	nodeInternalIPIndex = "nodeInternalIPIndex"
	nodeProviderIDIndex = "nodeProviderIDIndex"
)
...
if err := mgr.GetCache().IndexField(&corev1.Node{}, nodeProviderIDIndex, indexNodeByProviderID); err != nil {
	return nil, fmt.Errorf("error setting index fields: %v", err)
}
if err := mgr.GetCache().IndexField(&mapiv1beta1.Machine{}, machineProviderIDIndex, indexMachineByProviderID); err != nil {
	return nil, fmt.Errorf("error setting index fields: %v", err)
}
if err := mgr.GetCache().IndexField(&corev1.Node{}, nodeInternalIPIndex, indexNodeByInternalIP); err != nil {
	return nil, fmt.Errorf("error setting index fields: %v", err)
}
if err := mgr.GetCache().IndexField(&mapiv1beta1.Machine{}, machineInternalIPIndex, indexMachineByInternalIP); err != nil {
	return nil, fmt.Errorf("error setting index fields: %v", err)
}

```

Reply:
Done and created unit test for each indexer function